adapters/criteo/models.go

```diff
@@ -0,0 +1,333 @@
+		}
+	}
+
+	if site != nil {
+		if site.ID != "" {
+			criteoPublisher.SiteID = site.ID
+		}
+		if site.Page != "" {
+			criteoPublisher.URL = site.Page
+		}
+	}
+
+	return criteoPublisher
+}
+
+type criteoUser struct {
+	DeviceID     string `json:"deviceid,omitempty"`
+	DeviceOS     string `json:"deviceos,omitempty"`
+	DeviceIDType string `json:"deviceidtype,omitempty"`
+	CookieID     string `json:"cookieuid,omitempty"`
+	UID          string `json:"uid,omitempty"`
+	IP           string `json:"ip,omitempty"`
+	IPv6         string `json:"ipv6,omitempty"`
+	UA           string `json:"ua,omitempty"`
+	UspIab       string `json:"uspIab,omitempty"`
+}
+
+func newCriteoUser(user *openrtb2.User, device *openrtb2.Device, regsExt *openrtb_ext.ExtRegs) criteoUser {
+	criteoUser := criteoUser{}
+
+	if user == nil && device == nil {
+		return criteoUser
+	}
+
+	if user != nil {
+		if user.BuyerUID != "" {
+			criteoUser.CookieID = user.BuyerUID
+		}
+	}
+
+	if device != nil {
+		deviceType := getDeviceType(device.OS)
+		criteoUser.DeviceIDType = deviceType
+
+		if device.OS != "" {
+			criteoUser.DeviceOS = device.OS
+		}
+		if device.IFA != "" {
+			criteoUser.DeviceID = device.IFA
+		}
+		if device.IP != "" {
+			criteoUser.IP = device.IP
+		}
+		if device.IPv6 != "" {
+			criteoUser.IPv6 = device.IPv6
+		}
+		if device.UA != "" {
+			criteoUser.UA = device.UA
+		}
+	}
+
+	if regsExt != nil {
+		if regsExt.USPrivacy != "" {
+			criteoUser.UspIab = regsExt.USPrivacy // CCPA
+		}
+	}
+
+	return criteoUser
+}
+
+type criteoGdprConsent struct {
+	GdprApplies *bool  `json:"gdprapplies,omitempty"`
+	ConsentData string `json:"consentdata,omitempty"`
+}
+
+func newCriteoGdprConsent(user *openrtb2.User, regsExt *openrtb_ext.ExtRegs) (criteoGdprConsent, error) {
+	consent := criteoGdprConsent{}
+
+	if user == nil && regsExt == nil {
+		return consent, nil
+	}
+
+	if user != nil && user.Ext != nil {
+		var userExt *openrtb_ext.ExtUser
+		if err := json.Unmarshal(user.Ext, &userExt); err != nil {
+			return consent, err
+		}
+		if userExt.Consent != "" {
+			consent.ConsentData = userExt.Consent
+		}
+	}
+
+	if regsExt != nil {
+		if regsExt.GDPR != nil {
+			gdprApplies := bool((*regsExt.GDPR & 1) == 1)
+			consent.GdprApplies = &gdprApplies
+		}
+	}
+
+	return consent, nil
+}
+
+type criteoRequestSlot struct {
+	SlotID      string              `json:"slotid,omitempty"`
+	ImpID       string              `json:"impid,omitempty"`
+	ZoneID      *int64              `json:"zoneid,omitempty"`
+	NetworkID   *int64              `json:"networkid,omitempty"`
+	PlacementID string              `json:"placement,omitempty"`
+	Sizes       []criteoRequestSize `json:"sizes,omitempty"`
+}
+
+func newCriteoRequestSlots(slotIDGenerator slotIDGenerator, impressions []openrtb2.Imp) ([]criteoRequestSlot, []error) {
+	var errs []error
+
+	if len(impressions) == 0 {
+		return nil, []error{&errortypes.BadInput{
+			Message: "Bid request impressions is nil or empty",
+		}}
+	}
```

Comment:
There are no tests for this `len(impressions) == 0` condition. It's also not possible for the core logic of PBS to call `MakeRequests` for no impressions. Please consider either removing this check or adding a test.

Reply:
Ack